birdvoxpaint/birdvoxpaint.py

```diff
@@ -1,76 +1,110 @@
+from functools import partial
+
 import librosa
-from librosa.util.exceptions.ParameterError
+import librosa.display
 import numpy as np
 
+import tqdm
+import joblib
+
+from . import util
+from .indices import *
+from .display import *
+
+MAX_JOBS = 40
+
+def transform(filename=None, y=None, sr=None,
+              frame_length=2048, hop_length=512,
+
+              n_mels=256, fmin=None, fmax=None,
+              time_constant=0.4, use_pcen=True, use_mels=True,
+
+              indices=[average_energy],
+              segment_duration=10,
+              verbose=True, n_jobs=None, return_plot=False):
+    '''Convert audio into a false color spectrogram.
+
+    Arguments:
+        
+    '''
+
+    if not n_jobs or n_jobs < 0: # automatically use available cpus, limited by MAX_JOBS
+        n_jobs = min(joblib.cpu_count(), MAX_JOBS)
+
+    # load audio frames - generates one block at a time
+    y_blocks, sr = util.block_stream(
+        filename, y, sr,
+        frame_length=frame_length, hop_length=hop_length,
+        segment_duration=segment_duration, n_blocks=n_jobs) # yields: batch*time*n_fft
+
+    y_blocks = tqdm.tqdm(y_blocks, disable=not verbose)
+
+    # prepare spectrogram function
+    fft_state = {}
+    _fft_slice = util.freq_slice(fmin, fmax, sr, frame_length)
+    f_spec = partial(spec,
+        n_fft=frame_length, hop_length=hop_length,
+        win_length=frame_length, n_mels=n_mels,
+        sr=sr, fmin=fmin, fmax=fmax,
+        time_constant=time_constant,
+        use_mels=use_mels, use_pcen=use_pcen,
+         _fft_slice=_fft_slice, state=fft_state)
+
+    # convert audio blocks to spectrograms
+    S_blocks = (f_spec(y) for y in y_blocks) # yields: freq, time*batch
+
+    # calculate the number of frames in a segment (segment == block length / n_jobs)
+    segment_length = util.sample2frame(max(segment_duration * sr, frame_length), frame_length, hop_length)
+
+    # break spectrogram into blocks and apply indices
+    S = util.apply_indices(S_blocks, indices, segment_length, n_jobs=n_jobs)
+
+    if return_plot:
```

Comment:
i really don't think that we should have feature extraction and visualization be part of the same function. there's nothing wrong with having the user write two lines of code rather than one. and it considerably simplifies the functional pattern on our end. it is also more easily extensible.

Reply:
I just figured it would minimize errors when playing with parameters, but I guess if you set them at the top it wouldn't be a problem. easy to take out, it was just a convenience thing